include/open62541/util.h

```diff
@@ -68,6 +68,57 @@ UA_EXPORT UA_StatusCode
 UA_KeyValueMap_delete(UA_KeyValuePair **map, size_t *mapSize,
                       const UA_QualifiedName key);
 
+/**
+ * Create a deep copy of the given KeyValueMap. Function will allocate memory for
+ * the newly create KeyValueMap which needs to be freed by the user of the function
+ * accordingly.
+ * @param map The KeyValueMap to be copeid
+ * @param mapSize The size of the KeyValueMap
+ * @return A pointer to the copied KeyValueMap, NULL if memory allocation failed.
+ */
+UA_EXPORT UA_KeyValuePair*
+UA_KeyValueMap_copy(UA_KeyValuePair *map, size_t mapSize);
+
+/**
+ * Calculate the exact number of entries that will be needed for a merge of entries
+ * from two given KeyValueMaps.
+ * @param lhs The left hand side operand of the KeyValueMap merge
+ * @param lhsCount The count of entries in the lhs KeyValueMap
+ * @param rhs The right hand side operand of the KeyValueMap merge
+ * @param rhsCount The count of entries in the rhs KeyValueMap
+ * @return The exact number of entries needed for a merge of the two given KeyValueMaps
+ */
+UA_EXPORT size_t
+UA_KeyValueMap_countMergedMembers(UA_KeyValuePair *lhs, size_t lhsCount, UA_KeyValuePair *rhs, size_t rhsCount);
+
+/**
+ * Check KeyValueMap for containment
+ * ---------------------------------
+ * Check if a given KeyValueMap contains an entry mapped by the provided key.
+ * @param map The KeyValueMap to check against
+ * @param mapSize The size of the KeyValueMap
+ * @param key The key to check for existence in the KeyValueMap
+ * @return true if an entry with the given key exists, false otherwise
+ */
+UA_EXPORT UA_Boolean
+UA_KeyValueMap_contains(UA_KeyValuePair *map, size_t mapSize, UA_String key);
+
+/**
+ * Merge two KeyValueMaps
+ * ----------------------
+ * The two key value maps will be merged where values from the left operand
+ * will be overwritten by values of the right operand when the keys are identical.
+ *
+ * @param dst The KeyValueMap to store the resulting merge in. The size can either be an upper bound (rhsCout + lhsCount)
+ *            or can be calculated beforehand by using UA_KeyValueMap_countMergedMembers
+ * @param lhs The left hand side operand of the KeyValueMap merge
+ * @param lhsCount The count of entries in the lhs KeyValueMap
+ * @param rhs The right hand side operand of the KeyValueMap merge
+ * @param rhsCount The count of entries in the rhs KeyValueMap
+ */
+UA_EXPORT void
+UA_KeyValueMap_merge(UA_KeyValuePair *dst, UA_KeyValuePair *lhs, size_t lhsCount, UA_KeyValuePair *rhs, size_t rhsCount);
```

Comment:
This getting a bit too funky and complicated as an API.
There's a ton of edge cases that have to be handled when calling into _merge.

The KeyValueMap in general is still very new and not used in a release.
So we can still correct suboptimal design.

I would define
```c
struct UA_KeyValueMap {
    size_t size;
    UA_KeyValuePair  *map;
}
```
and then simplify to
```
UA_EXPORT UA_StatusCode
UA_KeyValueMap_merge(UA_KeyValueMap *lhs, const UA_KeyValuePair *rhs);
```

Note that the _set method already has "upsert" semantics.
So this is just a faster version of looping over the rhs and inserting everything.

Reply:
`KeyValuePair[]` is used throughout the standard whenever a function is accepting a variadic number of input parameters for specific things. E.g. the `ConnectionProperties` in the `PubSubConnectionType`.
I think the current ratio is, to have an array of KeyValuePair being a KeyValueMap and being able to use this KeyValueMap as input to the functional equivalents of such properties or parameters in the code.
If we on the other hand change the KeyValueMap to be its own actual type it will not have an equivalent in the OPCUA Type System (or will it?).
Anyway, the proposed way of merging via the `set` method sounds good. If one needs to keep an original `lhs` untouched, one can just use `copy` before.

The API of `countMerged` and `merge` was initially designed like that, because KeyValuePair arrays often where allocated on the stack, which would prevent an inplace enlargement of a KeyValuePair array.

Reply:
And of course, I also generally think it would be better to have an actual `UA_KeyValueMap` structure like you proposed, it would simplify a lot of places where we use pointer to KeyValuePair array and its size.

Reply:
Great. Then let's add the `UA_KeyValueMap` structure before the current API gets used in the wild.

> But with your proposed API, one would just have to make sure that the input arguments are heap allocated.

I think this information is generally known in places that would call _merge.